sale_coupon_multi_use/models/sale_order.py

```diff
@@ -1,32 +1,116 @@
-    # TODO: action_draft seems buggy, because it does not reset back
-    # coupons to new state (you can abuse coupon this way and use normal
-    # coupon multiple times).
+    coupon_multi_use_ids = fields.Many2many(
+        "sale.coupon",
+        "sale_order_coupon_multi_rel",
+        "sale_id",
+        "coupon_id",
+        string="Multi Use Coupons",
+        copy=False,
+        readonly=True,
+    )
+
+    def _get_multi_use_coupons(self):
+        self.ensure_one()
+        # NOTE. This method must be called with ORDER_CTX_KEY on order.
+        return self.coupon_multi_use_ids - self.applied_coupon_ids
 
     def action_confirm(self):
         """Extend to pass coupon_sale_order context."""
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            # Mimic same behavior as for single-user coupon.
+            order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_confirm()
+            order._get_multi_use_coupons().consume_coupons()
         return True
 
     def action_cancel(self):
         """Extend to pass coupon_sale_order context."""
-        # NOTE. Can't rely on coupon/SO relation, because
-        # applied_coupon_ids is o2m field, meaning same coupon used on
-        # another SO, would make lose relation on previous SO.
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            order = order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_cancel()
+            order._get_multi_use_coupons().reset_coupons()
+            # Mimic applied_coupon_ids logic.
+            order.coupon_multi_use_ids = [(5,)]
         return True
+
+    def _get_coupons_from_2many_commands(self, commands):
+        """Browse coupon IDs from 2many command."""
+        ids = []
+        for cmd in commands:
+            if cmd[0] == 6:
+                ids.extend(cmd[2])
+            elif cmd[0] == 4:
+                ids.append(cmd[1])
+        return self.env["sale.coupon"].browse(ids)
+
+    def _get_multi_use_coupons_from_2many_commands(self, commands):
+        coupons = self._get_coupons_from_2many_commands(commands)
+        return coupons.filtered("multi_use")
+
+    def write(self, vals):
+        """Extend to add multi-use coupons."""
+        if vals.get("applied_coupon_ids"):
+            # We want to add multi-use coupons that were added on
+            # standard applied_coupon_ids, so such coupon is preserved (
+            # applied_coupon_ids relation is removed once same coupon is
+            # used on multiple sale orders).
+            coupons_multi_use = self._get_multi_use_coupons_from_2many_commands(
+                vals["applied_coupon_ids"]
+            )
+            if coupons_multi_use:
+                vals["coupon_multi_use_ids"] = [(6, 0, coupons_multi_use.ids)]
+        return super().write(vals)
```

Comment:
For me all these 3 functions:
- _get_coupons_from_2many_commands
- _get_multi_use_coupons_from_2many_commands
- write

are too complex compared to the needs.

I think we can do more easy:
```
    def write(self, vals):
        """Extend to add multi-use coupons."""
        result = super().write(vals)
        if vals.get("applied_coupon_ids"):
            for order in self:
                order.coupon_multi_use_ids = order.applied_coupon_ids.filtered("multi_use")
        return result
```

Reply:
This calls write twice. Which is bad.

Reply:
It's not bad if the final result is very more easy, then more easy to read/maintainable.

Reply:
Hm, I checked your suggestion again and actually it would work incorrectly. It would cause same problem this PR was created to solve in a first place.

Thing is you need to check what is actually done with `applied_coupon_ids`. You can't assume correct command was used.

Now because `applied_coupon_ids` is o2m field, imagine this case:

SO1: we apply multi-use coupon on applied_coupon_ids (4 or 6 cmd used). It gets applied on coupon_multi_use_ids. So far so good.
Then we apply same multi-coupon on SO2 via applied_coupon_ids. Because that field its o2m field, relation will be lost on SO1.
Now let say we also apply single-use coupon on SO1. It gets applied on applied_coupon_ids and triggers part of code you suggested. Now it finds no multi-use coupons on applied_coupon_ids, thus overwrites all multi-use coupons and causes same problem we have without this PR.

Well I guess we could then only add new ones on coupon_multi_use_ids, but I wonder if then its any easier to read (though I dont really find my part hard to read, but thats my opinion:)).

Reply:
Ok I understand, then 2 things:

- Add all this explanation into the code to help people to correctly understand what you do and why you do it like this.
- `applied_coupon_ids` must always be empty when we have multi use module installed. Because the value of `applied_coupon_ids` is wrong, we don't want to keep it.

Reply:
applied_coupon_ids can't be kept empty, because it is also used for single use coupon (native functionality).
Though I guess what could be done, is to not apply multi-coupon IDs on `applied_coupon_ids` and instead of copying those ids, move them. So then it would simplify relation management (we would only check applied_coupon_ids once and on other parts deal with coupon_multi_use_ids)

Reply:
:+1: with your proposition! :)

Reply:
@jcoux updated. Moving 2many commands between keys got too complicated, so to refactored with multiple write. Though still needed to use commands there to make sure both fields are properly updated (because we need to be aware of specific 2many commands either way).